### ISSUE TITLE
2nd accelerate launch script impacts 1st scripts training speed, avoided by using launching via python script on 2nd script

### ISSUE THREAD

**Issue by minienglish1** [huggingface/accelerate#2632]
### System Info

```Shell
accelerate 0.28.0, ubuntu 22.04, torch 2.2.2

1) 1st script is a sdxl fine-tune, using FSDP, 2 rtx3090s (gpu_id 1,2)
2) 2nd script is cache latents to disk script, 1 rtx4090 (gpu_id 0).
The two scripts a completely independent and do not rely on or interact with each other in any way.

Running script #1 usually has speed ~7.3s/it.
If script is running script #2, script #1's speed decreases to ~9.4s/it.
	It is visible in nvtop that gpu_id 2 has increased idle time
If script 2 is modified to use python script2.py, instead of accelerate launch, no speed decrease occurs.

Attempts to fix the problem via setting main_process_port (assigning ports 29501/29502 or port 0) were unsuccessful

script #1

Accelerate config.yaml (FSDP_12.yaml):
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: FSDP
downcast_bf16: 'no'
fsdp_config:
  fsdp_auto_wrap_policy: SIZE_BASED_WRAP
  fsdp_backward_prefetch: BACKWARD_PRE
  fsdp_cpu_ram_efficient_loading: false
  fsdp_forward_prefetch: true
  fsdp_min_num_params: 50000000
  fsdp_offload_params: false
  fsdp_sharding_strategy: FULL_SHARD
  fsdp_state_dict_type: SHARDED_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_use_orig_params: true
gpu_ids: 1,2
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false

launch:
accelerate launch --config_file FSDP_12.yaml sdxl_FSDP_train_48.py

script #2

accelerate config.yaml (single_gpu_0.yaml)
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: 'NO'
downcast_bf16: 'no'
gpu_ids: '0'
machine_rank: 0
main_training_function: main
mixed_precision: fp16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false

launch:
accelerate launch --config_file single_gpu_0.yaml sdxl_process_data_dir_06.py 

python launch (no speed decrease in script #1):
python sdxl_process_data_dir_06.py
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

1) launch FSDP script with accelerate launch
2) launch a second single gpu non-distributed script with accelerate launch

There are no other in code reference to accelerate except for the: standard accelerator = Accelerator(gradient_accumulation_steps, mixed_precision; device = accelerator.device; & accelerator.prepare

### Expected behavior

2nd launched accelerate script should not affect 1st launched script's training speed.

**Comment by minienglish1** [comment]
I rewrote the 2nd script, removing all accelerate references from the script.  100% confirmed & reproducible that running a second accelerate script impacts 1st scripts training speed.  If using the 2nd script with no accelerate references, 1st script trains at exact same speed as only running 1st script.

I'd really like to help resolve this issue, as accelerate makes writing these scripts much easier.  I'll be happy to provide whatever information is needed.

**Comment by muellerzr** [comment]
How are you launching via python in the second script? Literally just `python myscript.py`? 

I don't have a 3-gpu system but I'll see if I can reproduce this on a cluster of 8, specifying 1 and 2 GPUs today.

My best *guess* is maybe CPU worker overlap?

**Comment by minienglish1** [comment]
yes, the second script is literally "python myscript.py" with an arg or two.
It's now literally just a regular python script, with no accelerate references inside.

I tried launching the 2nd script via python myscript.py (before having removed accelerate references), noticed initially what I thought was improvements, but as the script ran, the 1st script started slowing down.  So my comment above "If script 2 is modified to use python script2.py, instead of accelerate launch, no speed decrease occurs.", is not correct.  If the script has accelerator code, it causes a slow down.

My guess is also CPU worker overlap or is there some kind of main accelerate process in the background managing things and it's getting overloaded?  I tried adding more workers, but it mostly caused more problems.

I think it may also be related to another problem I'm having:
I'm using FSDP SHARD_GRAD_OP.  It works great with 2 GPUs,  When I add a 3rd gpu, it will slow down 1/3 the speed of 2 gpus.  Watching the gpu usage in nvtop, the gpu usage pattern looks very similar to what I am seeing with this 2 accelerate script issue: 1 gpu running at 100% with the other gpu(s) having short bursts of 0% usage.

**Comment by muellerzr** [comment]
cc @pacman100 maybe? Otherwise a non-accelerate issue 🤔

**Comment by minienglish1** [comment]
I've since installed another rtx4090, and can test running 2 separate multi-gpu accelerate scripts.

Running 2 separate multi-gpu accelerate scripts (1 on the 3090s, 1 on the 4090s), each with different main_process_port, appears to work fine.  There is a small slowdown, but to be expected due to increased bandwidth utilization.

What I don't understand, is the rtx4090 pair's training speed appears to be nearly random and is tied to each epoch.
I can run the script 10 times in a row, and get different training speeds each time.

I'm assuming the issue is somehow related to CPU workers or perhaps inter-gpu communications.

I've tried 1) ensuring no other processes are running on system 2) Adjusting process nice value (python & accelerate processes) 3) using taskset, all with no effect.  

Also increasing workers, (requires torch.multiprocessing.set_start_method('spawn') due to error), causes OOM errors.  I can't find a proper guide to resolve this issue, but I'll continue working on it.

In the meantime, I'm going to try some other methods to find a stable solution:
1)  accelerate FSDP FULL_SHARD, 
2) accelerate deepspeed zero stage 2 (previously OOMed with only 2 gpus, I'll try again since FSDP SHARD_GRAD_OP works with 2 gpus)
3) re-write the code using native pytorch FSDP


Thank you for all your help with this matter.  Even if I don't find a solution, I appreciate the time you spent looking into it.  Thanks again.

Attached is an example of the variable training speeds.  Each linear portion is an epoch.  I've never seen such wide variations, or tied to an epoch, training speeds in all of my experience training models.
![image](https://github.com/huggingface/accelerate/assets/124333820/3e058e6c-457c-44a9-9e92-4d54b6a7e120)
